### ISSUE TITLE
Battery information for linux with 'upower -d' command

### ISSUE THREAD

**Issue by BilalAM** [oshi/oshi#624]
In reference to this issue https://github.com/oshi/oshi/issues/306 , there has been a proposal to use the 'upower -d' command to extract the battery information instead of going through the files . This feature would be helpful as it doesnt require much calculations as most of the information is already provided or can be easily extracted like 'percentage remaining' or 'time to discharge'.

Sample output on my machine of upower -d gives me this 
``` 
Device: /org/freedesktop/UPower/devices/battery_BAT0
  native-path:          BAT0
  vendor:               Hewlett-Packard
  model:                Primary
  serial:               02281 2017/07/03
  power supply:         yes
  updated:              Wed 19 Sep 2018 11:58:18 PM PKT (109 seconds ago)
  has history:          yes
  has statistics:       yes
  battery
    present:             yes
    rechargeable:        yes
    state:               fully-charged
    warning-level:       none
    energy:              954.6 Wh
    energy-empty:        0 Wh
    energy-full:         957.234 Wh
    energy-full-design:  42.624 Wh
    energy-rate:         9.2056 W
    voltage:             15.384 V
    percentage:          100%
    capacity:            100%
    technology:          lithium-ion
    icon-name:          'battery-full-charged-symbolic''
```
Parsing this would be a bit difficult but it can be done.
I am doing this on my machine first then i'll be sharing the results.
The attributes I have selected to include are : 
-  Name
-  Model
-  Vendor
-  Remaining Percentage
-  Remaining Time

**Comment by dbwiddis** [comment]
What is in the output of `upower -d` that is not contained in one of the "files" in `/sys/class/power_supply/BAT0` or the output of `cat /sys/class/power_supply/BAT0/uevent` which OSHI uses?   Given that `/sys` is a virtual filesystem intended to directly access devices, it seems preferable/proper to use it rather than parsing a command line.

As for the 2241.5% capacity, can you give me that `uevent` file output so I can see where OSHI's coming up with that calculation?  (And compare it to `upower -d` while you're at it?)

**Comment by dbwiddis** [comment]
Looks like UPower [gets its info from sysfs](https://github.com/freedesktop/upower/tree/master/src/linux) anyway, although the individual files rather than the uevent summary.  Still should be no difference.

**Comment by BilalAM** [comment]
The weird capacity was due to an issue of my laptop battery itself (i discovered this last night ) , its not a bug or anything in the library.
As far as the upower -d command is concerned , I thought that instead of going through the uevent file and calculate the percentages and the time remaining , why not just parse that command to directly get the information that is already provided ?

**Comment by cilki** [comment]
Because command parsing is the least robust and usually least efficient method of obtaining info.

Also, OSHI should probably log a warning if the battery info retrieved from the system is invalid.

**Comment by BilalAM** [comment]
So if we want to extend the battery information , we need to go into individual files to get the required attributes like before?

**Comment by dbwiddis** [comment]
If I'm interpreting your question correctly -- yes, the preference is for OSHI to fetch data from native sources wherever possible.  For Linux, this includes the kernel interfaces designed into the `/proc` and `/sys` virtual filesystems.      I'm on the fence regarding single (virtual) file access (the current `uevent` fetching) vs. grabbing data from each individual (virtual) file (probably a more correct way to do it that eliminates parsing).   

But I'd love for someone to take on #306 if that's what you're asking.

@cilki I'm not so sure about a log warning.  Who decides what is "invalid"?   I'm in favor of directly passing on information from the OS to the user as accurately as possible, giving them the freedom to decide how to deal with unusual data rather than making the choice for them.

**Comment by cilki** [comment]
I was referring to the percentage of 2241.5 being invalid. Although I may have overlooked whether OSHI will calculate that or the user does.

In any case, filling up the logs with that kind of message may not be helpful anyway.

**Comment by dbwiddis** [comment]
Yep... but as I learned when capping CPU usage at 100% (and later realizing that in multi-threading a process CPU can exceed that) I'd rather give the user as much detail as possible and leave calculation to the user, or a util class.

For the battery, for example, I'm in favor of providing the user both "design capacity" and "current capacity" so they can determine for themselves that it's 2241.5%.   And have a util class with a "getCapacityPercent" method that will report the more sane 100%.

**Comment by dbwiddis** [comment]
I'm going to close this issue as #306 discusses most of this (including `upower -d` as an option).  If you have additional information that's not already included in #306 (like name/model/vendor), feel free to add that request to the other issue.
